source/src/microapp/cs_MicroappController.cpp

```diff
@@ -231,9 +234,20 @@ MicroappOperatingState MicroappController::getOperatingState(uint8_t appIndex) {
 
 void MicroappController::callMicroapp() {
 	const uint8_t appIndex = 0;
+
+	if (!Microapp::getInstance().canRunApp(appIndex)) {
+		return;
+	}
+
 	setOperatingState(appIndex, MicroappOperatingState::CS_MICROAPP_RUNNING);
+	uint32_t startTime = RTC::getCount();
 	if (resumeCoroutine()) {
+		uint32_t duration = RTC::msPassedSince(startTime);
 		setOperatingState(appIndex, MicroappOperatingState::CS_MICROAPP_NOT_RUNNING);
+
+		if (duration > MICROAPP_MAX_CALL_DURATION_MS) {
+			Microapp::getInstance().onExcessiveCallDuration(appIndex);
+		}
 		return;
 	}
```

Comment:
the end of this function: set relevant microapp state (e.g. started/running = false) to prevent jumping back in

Reply:
That's done by the microapp class, right?